File: scala/Problem37.scala
/*
The number 3797 has an interesting property. Being prime itself, it is possible
to continuously remove digits from left to right, and remain prime at each
stage: 3797, 797, 97, and 7. Similarly we can work from right to left: 3797,
379, 37, and 3.

Find the sum of the only eleven primes that are both truncatable from left to
right and right to left.

NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
*/

object Problem37 {
  def trunc(n:Int) = {
    val s = n.toString
    (s.tails.toList.init ++ s.inits.toList.tail).filter(""!=).map(_.toInt)
  }

  def isPrime(n:Int) = BigInt(n).isProbablePrime(10)

  lazy val candidates = (9 to 999999 by 2).filter(x => (x.toString.head.asDigit, x.toString.last.asDigit) match {
      case _ if(x<100) => true
      case (0|2|4|6|8, _) => false
      case (_, 0|2|4|6|8) => false
      case _ => true
  })

  def main(args:Array[String]) {
    print(candidates.filter(trunc(_).forall(isPrime)).take(11).sum)
  }
}

File: javascript/Problem37.js
/*
 * The number 3797 has an interesting property. Being prime itself, it is possible
 * to continuously remove digits from left to right, and remain prime at each
 * stage: 3797, 797, 97, and 7. Similarly we can work from right to left: 3797,
 * 379, 37, and 3.
 * 
 * Find the sum of the only eleven primes that are both truncatable from left to
 * right and right to left.
 * 
 * NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.
 */

function isPrime(n) {
	if (n < 2) {
		return false
	}
	if (n === 2) {
		return true
	}
  const limit = n ** 0.5
  for (let i = 2; i <= limit; i++) {
    if (n % i === 0) {
      return false
    }
  }
  return true
}

function trunc(n) {
  const digits = Array.from(String(n), Number)
  const ltor = Array
    .from({ length: digits.length - 1 }, (_, i) => i + 1)
    .map(x => digits.slice(0, x))
	digits.reverse()
	const rtol = Array
    .from({ length: digits.length - 1 }, (_, i) => i + 1)
    .map(x => digits.slice(0, x).reverse(false))
	return ltor
    .concat(rtol)
    .map(x => Number(x.join('')))
}

const result = []
for (let i = 9; result.length < 11; i += 2) {
  if (isPrime(i) && trunc(i).every(isPrime)) {
    result.push(i)
  }
}
const answer = result.reduce((x, y) => x + y)
console.log(answer)
